parlai/scripts/convo_render.py

```diff
@@ -0,0 +1,241 @@
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import os
+import sys
+import json
+import random
+import argparse
+import tempfile
+
+# Constant Denoting the End of a Conversation
+END_OF_CONVO = "EOC"
+
+
+def gen_ul_style():
+    """ Generate general ul tag style"""
+    return (
+        "\t\tul{\n\t\t\tlist-style: none;\n\t\t\tmargin: 0;\n\t\t\tpadding: 0;\n\t\t}\n"
+    )
+
+
+def gen_ul_li_style():
+    """ Generate style for the li items for the conversations"""
+    return "\t\tul li{\n\t\t\tdisplay:inline-block;\n\t\t\tclear: both;\n\t\t\tpadding: 20px;\n\t\t\tborder-radius: 30px;\n\t\t\tmargin-bottom: 2px;\n\t\t\tfont-family: Helvetica, Arial, sans-serif;\n\t\t}\n"
+
+
+def gen_breaker_style():
+    """ Generate style for the break li item between conversations"""
+    return "\t\t.breaker{\n\t\t\tcolor: #bec3c9;\n\t\t\tdisplay: block;\n\t\t\theight: 20px;\n\t\t\tmargin: 20px 20px 20px 20px;\n\t\t\ttext-align: center;\n\t\t\ttext-transform: uppercase;\n\t\t}\n"
+
+
+def gen_speaker_styles(other_speaker, human_speaker):
+    """
+    Generate style string for the speakers
+    :param other_speaker: The title of the model (grey boxes)
+    :param human_speaker: Human speaker in the dialogs (blue boxes)
+
+    :return: Style string for the speakers
+    """
+    other_style = (
+        "\t\t."
+        + other_speaker
+        + "{\n\t\t\tbackground: #eee;\n\t\t\tfloat: left;\n\t\t}\n"
+    )
+    human_style = (
+        "\t\t."
+        + human_speaker
+        + "{\n\t\t\tbackground: #0084ff;\n\t\t\tcolor: #fff;\n\t\t\tfloat: right;\n\t\t}\n"
+    )
+
+    return other_style + human_style
+
+
+def gen_style_str(height, width, other_speaker, human_speaker):
+    """
+    Generate the style section string of the HTML
+    :param height: Height of the HTML page
+    :param width: Width of the HTML page
+    :param other_speaker: The title of the model (grey boxes)
+    :param human_speaker: Human speaker in the dialogs (blue boxes)
+
+    :return: The <style> section string
+    """
+    ul_style = gen_ul_style()
+    ul_li_style = gen_ul_li_style()
+    speaker_styles = gen_speaker_styles(other_speaker, human_speaker)
+    breaker_style = gen_breaker_style()
+    style_str = (
+        '\t<style type="text/css">\n\t\t@media print {\n\t\t\t@page { margin: 0; size:'
+        + str(width)
+        + "in "
+        + str(height)
+        + "in;}\n\t\t}\n"
+        + ul_style
+        + ul_li_style
+        + speaker_styles
+        + breaker_style
+        + "\n\t</style>\n"
+    )
+
+    return style_str
+
+
+def gen_convo_ul(conversations):
+    """
+    Generate the ul section of the HTML for the conversations
+    :param conversation: The conversation to be rendered (after pre-processing)
+
+    :return: The string generating the list in HTML
+    """
+    ul_str = "\t<ul>\n"
+    for speaker, speech in conversations:
+        if speaker == END_OF_CONVO:
```

Comment:
I don't think we need "End of Conversation" markers, they'll be implied by actually reaching the end, no?

Reply:
For CLI yes, but when rendering HTML, we pass our function simply a collection of speeches, so there isn't a way to track where a conversation ends and where a new one begins. I thought its nice to track that since in the HTML I mark end of conversation in a similar manner as what actually happens in messenger (except that there a date is used)

Reply:
`<hr/>`?

Reply:
Tried that alternative, doesn't look as nice as the current solution (looks more like messenger right now), but if you prefer that I can change to that.

Reply:
Yes please, or something along these lines.

Reply:
Done in the latest commit!